tests/test_registries.py

```diff
@@ -0,0 +1,166 @@
+
+    @base.addfixture(fixtures.EnvironmentVariable, 'DOCKER_USER', 'my_user')
+    @base.addfixture(fixtures.EnvironmentVariable, 'DOCKER_PASSWORD',
+                     'my_password')
+    def test_credentials_ignore_env_when_set(self):
+        registry = windlass.registries.DockerRegistry(
+            self.reg_url,
+            "explicit_user",
+            "explicit_password",
+        )
+
+        self.assertEqual('explicit_user', registry.username)
+        self.assertEqual('explicit_password', registry.password)
+
+
+@unittest.mock.patch('windlass.registries.remotes.ECRConnector')
+class TestEcrRegistry(testtools.TestCase):
+
+    reg_url = "000000000000.dkr.ecr.us-west-1.amazonaws.com"
+
+    @base.addfixture(fixtures.EnvironmentVariable, 'AWS_ACCESS_KEY_ID',
+                     'AKIAIOSFODNN7EXAMPLE')
+    @base.addfixture(fixtures.EnvironmentVariable, 'AWS_SECRET_ACCESS_KEY',
+                     'wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY')
+    @base.addfixture(fixtures.EnvironmentVariable, 'AWS_DEFAULT_REGION',
+                     'us-west-2')
+    def test_credentials_from_env(self, ecr_connector_mock):
+        registry = windlass.registries.EcrDockerRegistry(self.reg_url)
+
+        self.assertIsNone(registry.username)
+        self.assertIsNone(registry.password)
+        call_args = ecr_connector_mock.call_args_list
+        # check that ECRConnector class is called with AWSCreds containing
+        # the example account key id as the first arg
+        self.assertEqual(call_args[0][0][0][0], 'AKIAIOSFODNN7EXAMPLE')
+        self.assertEqual(call_args[0][0][0][2], 'us-west-2')
+
+    @base.addfixture(fixtures.EnvironmentVariable, 'AWS_ACCESS_KEY_ID',
+                     'AKIAIOSFODNN7EXAMPLE')
+    @base.addfixture(fixtures.EnvironmentVariable, 'AWS_SECRET_ACCESS_KEY',
+                     'wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY')
+    def test_credentials_default_region_from_url(self, ecr_connector_mock):
+        registry = windlass.registries.EcrDockerRegistry(self.reg_url)
+
+        self.assertIsNone(registry.username)
+        self.assertIsNone(registry.password)
+        call_args = ecr_connector_mock.call_args_list
+        # check that ECRConnector class is called with AWSCreds containing
+        # the extracted region from the url
+        self.assertEqual(call_args[0][0][0][2], 'us-west-1')
+
+    @base.addfixture(fixtures.EnvironmentVariable, 'DOCKER_USER', 'my_user')
+    @base.addfixture(fixtures.EnvironmentVariable, 'DOCKER_PASSWORD',
+                     'my_password')
+    @base.addfixture(fixtures.EnvironmentVariable, 'AWS_SECRET_ACCESS_KEY', '')
+    def test_credentials_fallback_to_docker(self, ecr_connector_mock):
+        registry = windlass.registries.EcrDockerRegistry(self.reg_url)
+
+        self.assertEqual('my_user', registry.username)
+        self.assertEqual('my_password', registry.password)
+
+
+class TestJfrogRegistry(testtools.TestCase):
```

Comment:
Similar move to "useFixture" could happen here, perhaps with having two classes.